### ISSUE TITLE
[BUG] 4.2.0 doesn't work with styled-components

### ISSUE THREAD

**Issue by L3ARNING2B** [reactstrap/reactstrap#355]
Hi everyone,

I have read this very recent [article](https://www.smashingmagazine.com/2017/01/styled-components-enforcing-best-practices-component-based-systems/) claiming styled-components is the future so I decided to give a try.

I want to to customize my [reactstrap](https://reactstrap.github.io/) app. _(React equivalent of Bootstrap 4)_ 

I don't wan't to import the bootstrap css stylesheet. I rather make my own styling based on the class used in bootstrap. This is where styled-components come from! I Hoped...

For some reason, the custom theme passed to the `<ThemeProvider theme={theme} />` disappear for the one I've configured in my component `Button.defaultProps` right after a re-rending due to state change.

I've looked into the [__Button__ components](https://github.com/reactstrap/reactstrap/blob/master/src/Button.js) and I still wonder why. Components look fairly simple.

I have created a webpackbin to demonstrate the issue.

## Version
styled-components 2.0.0-7
reactstrap 4.2.0

## Reproduction

https://www.webpackbin.com/bins/-KetejqQhRZrV5MyzPpr

## Steps to reproduce

Click on the button within the reproduction and see.

## Expected Behavior

Custom theme to be applied

## Actual Behavior

Default theme is applied

**Comment by L3ARNING2B** [comment]
I've opened an equivalent issue here https://github.com/styled-components/styled-components/issues/574

**Comment by TheSharpieOne** [comment]
Yeah, definitely looks like an issue on styled-components side. The button, and other reactstrap components just renders the classes which it is told to render. Looking at the example you gave, the `Styled(*)` component has an internal state `generatedClassName` which matches what the button is getting/using.
Also, it appears that the theme is making it's way to the props which are passed to the Button... "Warning: Unknown prop `theme` on <button> tag. Remove this prop from the element. For details, see https://fb.me/react-unknown-prop"... Seems like `Styled(*)` shouldn't be passing down the props it uses.
IDK, but looking at [this line in styled-components](https://github.com/styled-components/styled-components/blob/master/src/hoc/withTheme.js#L22) it looks like it only subscribes and doesn't set the initial state based on the [default] props.

**Comment by L3ARNING2B** [comment]
> Yeah, definitely looks like an issue on styled-components side.

This is what I am trying to determinate. Why does this issue now appear that I use reactstrap components.

>  Looking at the example you gave, the Styled(*) component has an internal state generatedClassName which matches what the button is getting/using.

This part work fine, reactstrap doesn't exclude the passed className this is why you can still have the `[default] props.theme` at every moment.

> Also, it appears that the theme is making it's way to the props which are passed to the Button... "Warning: Unknown prop theme on tag. Remove this prop from the element. For details, see https://fb.me/react-unknown-prop"...

This is because the `props.theme` should be omitted before passing to the `<Button />` component. This can be ignored and is not related to the issue.

> Seems like Styled(*) shouldn't be passing down the props it uses.

There is two options:

1. Add `[default] props.theme` to your Component
1. Use a `<ThemeProvider theme={{ green: 'green' }}>`that will follow the theme to it's children nodes through their props.

> IDK, but looking at this line in styled-components it looks like it only subscribes and doesn't set the initial state based on the [default] props

Yes, I have the feeling this subscription is especially for the user theme passed to the ThemeProvider.
It get broken and restored If I switch to another page.

__styled-components__ doesn't have a function that return a `<ThemeProvider />` with a user theme in it's props.
I haven't though about that Idea but it might work don't you think?

**Comment by TheSharpieOne** [comment]
I only looked at their codebase for 60 seconds so it's hard for me to tell. As far as reactstrap is concerned, reactstrap is being passed a className and using it reactstrap using it.

**Comment by santaclauze** [comment]
I have managed to fix the issue but I am not certain on how to behave for my future code.

# Version

styled-components 2.0.0-7
reactstrap 4.2.0

# Steps to reproduce

I have created 3 cases:

1. __A__: Component created with `styled.button`
1. __B__: Component created with `class Button2 extends React.Component {}` with external _[default] props_
1. __C__: Component created with `class Button3 extends React.Component {}` with internal _static [default] props_

Case B use [the official recommended syntax for writing props](https://facebook.github.io/react/docs/typechecking-with-proptypes.html#default-prop-values)

Case C use `static` keyword within the class from [the proposition of babel](http://babeljs.io/blog/2015/06/07/react-on-es6-plus#property-initializers), __Webpackbin does not support the `static` keyword so I couldn't reproduce it in the bin__.
Copy Paste example and uncomment lines 11 and 13.

- In case A: within styles
- In case B and C: within render and within styles

https://www.webpackbin.com/bins/-Kf5Tdg7bfj2oWZzIb1V

Problematic Case:
```jsx
const defaultProps = {
  theme: themeBs,
};

class ButtonClass extends React.Component {
...
}

ButtonClass.defaultProps = defaultProps;
export default ButtonClass;
```

Fixed Case (Case C):

```jsx
const defaultProps = {
  theme: themeBs,
};

class ButtonBetter extends React.Component {
  static defaultProps = defaultProps;
...
}

export default ButtonBetter;
```

Moving the defaultProps inside the component using static keyword solve the issue.

# Expected Behavior

Custom theme to be applied

# Actual Behavior

Default theme is applied

# Questions

1. Is there any way to use ` static` keyword in Webpackbin? Or is there any work around to produce an equivalent results? In other words, what does the static keyword exactly do and how to reproduce it without him?

1. It seems the class doesn't contain publicly  `Component.defaultProps` or `Component.staticProps` anymore. Should we support the react way or the babel way ? 

1. What will happen to stateless components and styled if we are obliged to define `defaultProps` within the component? (As it seems that you cannot define static in non class component (`styled.button`, stateless components, ...))

**Comment by santaclauze** [comment]
We actually figured out the answer which was unrelated to the ReactStrap code. However we did post the question here as it may have had incidence on the reactstrap code.

Answers are as follows:

1. You can deactivate eslint on webpackbin in settings, ther eis no other support at this time for this issue.

1. Both different codes are equivalent, that is Problematic Case = to Fixed Case (as described above), [see proof] (https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Creact%2Cstage-2&targets=&browsers=&builtIns=false&code=class%20A%20%7B%0A%09static%20a%20%3D%20'hi'%0A%7D%0A%0Aconsole.log(A.a)%0A%0Aclass%20B%20%7B%0A%7D%0A%0AB.b%20%3D%20'hi'%0A%0Aconsole.log(B.b)).

Our issue was related to the sequence of code, where the attaching default props before using styled would apply default props after styling therefor creating issues with final result.

**Comment by eddywashere** [comment]
Are these questions best suited for an issue in styled-components library?

**Comment by santaclauze** [comment]
I believe the question was asked here as originally we were not sure where the issue was coming from and if it was coming from ReacStrap then it was worth asking here as well (as we posted the question on the other library). 

I did not close the topic, because I am unable to, but I answered, in the same line as you, that the issue does not lay within this library. 

But thanks for your concern! :)
